### DISCUSSION TITLE
Plan to Upgrade OBS Studio to Qt 6

### DISCUSSION THREAD

**Discussion by RytoEX** [obsproject/obs-studio#6481] [Announcements]
# Summary

As the title says, we're planning to move from Qt 5 to Qt 6 for OBS Studio 28. Over the past year, we've made changes to make the OBS codebase compatible with both Qt 5 and Qt 6, and we've made build system changes to more easily allow us to build against either Qt 5 or Qt 6. With Qt 6 now up to Qt 6.3, and Qt 6 being readily available on [Flatpak](https://invent.kde.org/packaging/flatpak-kde-runtime/-/tree/qt6.3), [Debian](https://tracker.debian.org/pkg/qt6-base), and now [Ubuntu 22.04](https://packages.ubuntu.com/jammy/qt6-base-dev) we feel the time is right to make the upgrade.

# What Version of Qt?

We plan to use the following Qt versions per platform:

| Platform            | Qt Version |
| ------------------- | ---------- |
| Windows             | 6.3        |
| macOS               | 6.3        |
| Flatpak             | 6.3        |
| Ubuntu 22.04+ (PPA) | 6.2        |
| Ubuntu 20.04 (PPA)  | 5.12       |

We will still ship OBS Studio 28 via PPA for Ubuntu 20.04 built against Qt 5.12.x. We will be dropping support for Ubuntu 18.04.

# When Will This Happen?

We're planning to enter the beta cycle of OBS Studio 28 in early-mid June. We're planning to have OBS Studio 28 fully released around the end of August.

# Why Upgrade?

We currently backport a number of patches into Qt 5.15 that we would not have to maintain if we move to Qt 6. Qt 6 also provides better support for HiDPI and fractional scaling, as well as fixes for a number of QTBUGs that we are tracking and additional improvements.

# Plugins

Plugins that link directly against Qt will have to be compiled against Qt 6 to be compatible with OBS Studio 28. This means that if a plugin author whose plugin links directly against Qt wishes to support both OBS Studio 28 and older versions of OBS Studio, they will have to make separate releases for OBS Studio 28+ and versions older than OBS Studio 28, linked against Qt 6 and Qt 5 respectively.

Plugin developers who wish to continue to support Ubuntu 20.04 natively will have to make their plugin source code compatible with Qt 5 at minimum.

If you need any help with this, please feel free to reach out to me here, in the [OBS Discord server](https://obsproject.com/discord), or on Discord directly (RytoEX#5545).

# 32-bit Support

Qt 6 does not support 32-bit builds. As such, we will no longer provide 32-bit builds of OBS as of OBS Studio 28. If you rely on tools that currently require 32-bit OBS, you may wish to contact the providers of those tools to see if they are able to provide 64-bit versions.

Please note that 64-bit builds of OBS are able to capture 32-bit games, and 32-bit applications are able to use the OBS Virtual Camera from a 64-bit OBS.

# Platform Support

Qt 6 no longer supports the following platforms:
* Windows 7
* Windows 8
* Windows 8.1
* macOS 10.13
* Ubuntu 18.04

As such, we cannot guarantee that OBS Studio 28 and newer will run on these platforms.

Furthermore, [Qt 6.3 only fully supports the following Windows and macOS versions](https://doc.qt.io/qt-6.3/supported-platforms.html):
* macOS 10.14, 10.15, 11, 12
* Windows 11 (21H2 or later)
* Windows 10 (21H2 or later)

We cannot guarantee that OBS will run well or at all on Windows or macOS versions older than these.

**Comment by derrod** [comment]
FWIW, last Steam Hardware Survey recorded the following figures for pre-Win 10 and 32-bit Windows users:
| Category / User Percentage | Global | United States | Germany | China | Russia
|----------------------------|--------|---------------|---------|-------|-------|
| 32-bit Windows             | 0.24%  | 0.03%         | 0.1%    | 0.19% | 0.46% |
| Windows 7, 8, 8.1          | 3.94%  | 0.6%          | 0.9%    | 7.5%  | 6.5% |

See: https://store.steampowered.com/hwsurvey/ for Global stats (country-specific ones are only available to Partners afaik).

**Comment by RytoEX** [comment]
With https://github.com/obsproject/obs-studio/pull/6877 merged, the OBS Studio CI builds for Windows and macOS are now using Qt 6. Similarly, the now released [OBS Studio 28.0 Beta 1](https://github.com/obsproject/obs-studio/releases/tag/28.0.0-beta1) has also shipped with Qt 6. This means that the builds available via our GitHub Actions CI and the OBS Studio 28 beta releases will no longer work with plugins that link against Qt 5. Those plugins will have to update to Qt 6 to continue working with these builds and the upcoming OBS Studio 28 release.

As before, if you need any help with this, please feel free to reach out to me here, in the [OBS Discord server](https://obsproject.com/discord), or on Discord directly (RytoEX#5545).

Here are some pull requests that address some Qt deprecations that may be helpful for updating your plugins to be compatible with Qt 5 and Qt 6:
* https://github.com/obsproject/obs-studio/pull/3810
* https://github.com/obsproject/obs-studio/pull/3811
* https://github.com/obsproject/obs-studio/pull/4299
* https://github.com/obsproject/obs-studio/pull/4318
* https://github.com/obsproject/obs-studio/pull/4352
* https://github.com/obsproject/obs-studio/pull/4358
* https://github.com/obsproject/obs-studio/pull/5254
* https://github.com/obsproject/obs-studio/pull/5654
* https://github.com/obsproject/obs-studio/pull/6456
* https://github.com/obsproject/obs-studio/pull/6672
* https://github.com/obsproject/obs-studio/pull/6770
* https://github.com/obsproject/obs-studio/pull/6788

The [obs-plugintemplate](https://github.com/obsproject/obs-plugintemplate) repo is also a good resource.

**Comment by theofficialgman** [comment]
I am just commenting here to stop the incorrect statements that I see going about on many projects. QT6 is NOT incompatible with bionic, not in source. QT offers binary packages for x86_64 built with a dependency on GLIBC 2.28 (so they are probably using debian buster to build now) and that can not be run on bionic.
but if you are building from source (which is what all the PPAs do) QT6 it is compatible with bionic. There are already multiple PPAs, example of one is here: https://launchpad.net/~okirby/+archive/ubuntu/qt6-backports/+packages

**Reply by RytoEX** [reply]
What I wrote was:
> Qt 6 no longer supports the following platforms:
>  * Windows 7
>  * Windows 8
>  * Windows 8.1
>  * macOS 10.13
>  * Ubuntu 18.04

I did not claim that Qt 6 is not compatible with Ubuntu 18.04.

[Qt 6's platform support page](https://doc.qt.io/qt-6.2/supported-platforms.html), which I previously linked in the original post, does not list Ubuntu 18.04. They simply do not officially support the platform, regardless of whether or not it can be built for or run on that platform. Coincidentally, we were dropping 18.04 _independently_ as a result of other changes we had made to obs-studio as well as the fact that we typically only support back to the latest or second latest available Ubuntu LTS release. Given that both Ubuntu 20.04 and 22.04 are available, we decided that dropping Ubuntu 18.04 in tandem with the upgrade to Qt 6, along with other aforementioned reasons, was the right time to do so.

Its availability as a third-party backport is of no consequence to our decision on this matter.

**Comment by juanpc2018** [comment]
Even if Windows becomes mandatory like Vaccines, i prefer to die. LOL.™
